README.rst
An Email Throttler for Django
=============================

Django error logging and emailing on a busy website can send a *lot* of emails,
all of them containing large traces. This can quickly become a bottleneck.

However, if there's an actual problem, most of the mails just repeat what's
already reported, so there's not much point in sending them all; a "sample" is
enough. It's useful to filter out repeating ones. This package does exactly
that.

It is possible to use this for something else than error-report-throttling,
but since the purpose is to hold back some emails, it may hurt you more than
it helps.


How Does It Work?
-----------------

The package contains a Django email backend that has a local cache of what
emails were sent (based on subject, not content). If the total amount of mails
per time unit goes over an overall threshold, then the excess is just counted,
not sent out. Similarly, if the amount of mails with the same subject line is
over a threshold, then the excess is just counted, not sent out. Both
thresholds are optional and they can be different.

The mails that are not throttled are sent out with the regular Django SMTP
backend. They have an additional "X-Mail-Throttler" header which also records
bits of statistics about the current values for threshold counters.

The package includes a management command that can be run from cron. If there
are any throttled emails, then this will provide a report including statistics
about how many mails have been held back.

The cache is maintained in files on the filesystem. The method used is highly
efficient for this purpose, and doing something similar in a database could
quickly lead to performance issues which will make the overload worse, not
better.


Installation and Configuration
------------------------------

Install the package with pip, for example:

.. code:: bash

	pip install django-email-throttler

You need to add a reference in the **INSTALLED_APPS** section in the Django **settings.py** file:

.. code:: python

	INSTALLED_APPS = (
		# ... whatever other apps ...
		'django_email_throttler',
	)

Settings and Defaults
~~~~~~~~~~~~~~~~~~~~~

Add these to your settings:

.. code:: bash

    # stuff for email throttling and their defaults
    EMAILTHROTTLER_TMPDIR = '/tmp'
    EMAILTHROTTLER_PREFIX = 'emailthrottler-'
    EMAILTHROTTLER_INTERVAL = 600
    EMAILTHROTTLER_TRUNCATE_SUBJECT = None # 128 is also ok
    EMAILTHROTTLER_SUBJECT_THRESHOLD = 3 # 0 for no-check
    EMAILTHROTTLER_OVERALL_THRESHOLD = 15 # 0 for no-check

    LOGGING = {
        'version': 1,
        'disable_existing_loggers': True,
        'handlers': {
            'mail_admins': {
                'level': 'ERROR',
                'class': 'django.utils.log.AdminEmailHandler',
                'email_backend': 'django_email_throttler.throttler.ThrottledEmailBackend',
                'include_html': True,
            }
        },
        'loggers': {
            'django.request': {
                'handlers': ['mail_admins'],
                'level': 'ERROR'
            },
        }
    }


Reporting Configuration
~~~~~~~~~~~~~~~~~~~~~~~

You should add a crontab entry to mail out the statistics and clean the
database. Make sure the frequency you're using to run this with is aligned to
the intervals you set. Something like this would work, assuming you have your
virtualenv kick in automagically:

.. code:: bash

    */10 * * * * $WHATEVER_PATH/manage.py throttle_report



CHANGES.rst
Release History
===============

0.5.0 (2023-08-25)
------------------

- Django 4.1 compatibility change (thanks @l00sed!)
- Drop Python 2.x support

0.4.1 (2019-08-21)
------------------

- fix syntax error (thanks @billrideout!)

0.4.0 (2019-08-07)
------------------

- Python3 support (thanks @chrisamin!)


0.3.4 (2018-04-19)
------------------

- Added EMAILTHROTTLER_TRUNCATE_SUBJECT (integer) to configuration to limit
  the length of file names created by the middleware. This can come handy if
  you hit file system limitations. Default value is None.


0.3.3 (2016-09-20)
------------------

- Use "requires_system_checks = False" to disable system checks on startup.
  It comes handy e.g. when the main database in unavailable: this way the
  reporting still works.


0.3.2 (2016-08-18)
------------------

- change optparse to argparse, which is enforced by Django 1.10


0.3.1 (2016-01-26)
------------------

- pypi was acting up, bumping version to re-publish


0.3.0 (2016-01-26)
------------------

- use OSError instead of FileNotFoundError for python 2 compatibility


0.2.7 (2016-01-26)
------------------

- encode email subjects as urlsafe_base64 to avoid filename problems


0.2.6 (2016-01-24)
------------------

- clean up pypi packaging


0.2.5 (2016-01-18)
------------------

- Change to a proper app name: django_email_throttler


0.2.3 (2016-01-18)
------------------

- Various changes to make a pypi package


0.2.1 (2016-01-18)
------------------

- Adding a MANIFEST.in


0.2.0 (2016-01-18)
------------------

- Applied some wisdom from @searchingfortao
- Remove dependency on arrow at the expense of having to use ISO8601 when
  specifying the --start-date option


0.1.0 (2016-01-15)
------------------

Initial release
~~~~~~~~~~~~~~~

- support for file based bookkeeping
- support for per subject and overall limits
- support for variable bucketing
- support for reporting or not, log cleanup or not


django_email_throttler/management/commands/throttle_report.py
import datetime
import dateutil.parser
import os
import glob
from base64 import urlsafe_b64decode

from django.core.mail import mail_admins
from django.core.management.base import BaseCommand

from ...throttler import ThrottledEmailBackend


class Command(BaseCommand):
    """
        Make a report about throttled mails and/or clean up the database.
        This command should be run once per throttle interval, preferably
        not much after the interval is over, most likely from cron.
        For example, if your interval is the default 600 seconds (10 minutes),
        then run it from cron as "*/10 * * * *" or so
    """

    help = "Report on email throttling"
    requires_system_checks = []

    def add_arguments(self, parser):

        parser.add_argument(
            '--no-clean-log',
            action='store_true',
            dest='no_clean_log',
            default=False,
            help="Keep email log"
        )

        parser.add_argument(
            '--force-email',
            action='store_true',
            dest='force_email',
            default=False,
            help="Send summary email even if there was no throttling (but only if there were emails in general)"
        )

        parser.add_argument(
            '--no-email',
            action='store_true',
            dest='no_email',
            default=False,
            help="Don't sent summary email even if there would be something to report"
        )

        parser.add_argument(
            "--start-date",
            action="store",
            dest="start_date",
            default=False,
            help="Use this as the starting time of an interval-to-check. Format is 'YYYY-MM-DD HH:MM' (UTC)"
        )

    # command handler
    def handle(self, *args, **options):

        # define our time interval
        if options['start_date']:
            bb = dateutil.parser.parse(options['start_date'])
        else:
            bb = self._bucket_begin()
        be = self._bucket_end(bb)

        # report if there are mails and either there's something to report or report is forced
        if not options['no_email']:
            self._send_report(bb, be, options['force_email'])

        if not options['no_clean_log']:
            self._clean_log(bb)

    # assemble and send email report
    @classmethod
    def _send_report(cls, bb, be, force_email):

        n_no, n_subject, n_overall, details = cls._collect_data(bb)

        subject = "Mail throttling report for {bb} -- T:{t} N:{n} S:{s} O:{o}".format(
            bb=bb.strftime('%Y-%m-%d %H:%M'),
            t=n_no+n_subject+n_overall,
            n=n_no,
            s=n_subject,
            o=n_overall,
        )

        out = ""
        out += "Report date: {bb} - {be}\n".format(bb=bb, be=be)
        out += "Total mails: {}\n".format(n_no+n_subject+n_overall)
        out += "Not throttled: {}\n".format(n_no)
        out += "Throttled on subject: {}\n".format(n_subject)
        out += "Throttled on overall amount: {}\n".format(n_overall)
        out += "\n"
        out += "T/N/S/O Subject\n"

        # dump the collection in a nice format
        for key, value in details.items():
            out += "{t}/{n}/{s}/{o}\t'{subject}'\n".format(
                t=value[ThrottledEmailBackend.THROTTLE_NO] +
                  value[ThrottledEmailBackend.THROTTLE_SUBJECT] +
                  value[ThrottledEmailBackend.THROTTLE_OVERALL],
                n=value[ThrottledEmailBackend.THROTTLE_NO],
                s=value[ThrottledEmailBackend.THROTTLE_SUBJECT],
                o=value[ThrottledEmailBackend.THROTTLE_OVERALL],
                subject=urlsafe_b64decode(key).decode('utf-8')
                )

        # send this report if there's something to say
        if (n_no+n_subject+n_overall > 0) and (force_email or n_subject > 0 or n_overall > 0):
            mail_admins(subject, out)

    # clean up the entries for the current interval from the db
    @classmethod
    def _clean_log(cls, bb):
        pattern = os.path.join(ThrottledEmailBackend.tmpdir_name,
                               ThrottledEmailBackend.file_prefix+str(bb)+"*")
        for f in glob.glob(pattern):
            if os.path.isfile(f):
                os.remove(f)

    # assuming that the script is run not much after the current bucket started
    # calculate start time of the *previous* bucket
    @classmethod
    def _bucket_begin(cls):
        epoch_now = int(datetime.datetime.utcnow().strftime('%s'))
        return datetime.datetime.fromtimestamp(
            epoch_now - (epoch_now % ThrottledEmailBackend.interval_size) - ThrottledEmailBackend.interval_size
        )

    # calculate end time of a bucket based on the start time
    @classmethod
    def _bucket_end(cls, bb):
        return datetime.datetime.fromtimestamp(
            int(bb.strftime('%s')) + ThrottledEmailBackend.interval_size
        )

    # get number of total, subject/overall filtered messages for a bucket
    # and all details too
    @classmethod
    def _collect_data(cls, bb):
        n_no = 0
        n_subject = 0
        n_overall = 0
        collector = {}

        pattern = os.path.join(ThrottledEmailBackend.tmpdir_name,
                               ThrottledEmailBackend.file_prefix + str(bb) + "*")
        for f in glob.glob(pattern):
            if os.path.isfile(f):
                subject = f[f.index('^')+1:]

                collector[subject] = {}
                with open(os.path.join(ThrottledEmailBackend.tmpdir_name, f),
                          "r") as subject_file:
                    content = subject_file.read()

                    # count
                    this_no = content.count(str(ThrottledEmailBackend.THROTTLE_NO))
                    this_subject = content.count(str(ThrottledEmailBackend.THROTTLE_SUBJECT))
                    this_overall = content.count(str(ThrottledEmailBackend.THROTTLE_OVERALL))

                    # store per subject and update totals
                    collector[subject][ThrottledEmailBackend.THROTTLE_NO] = this_no
                    collector[subject][ThrottledEmailBackend.THROTTLE_SUBJECT] = this_subject
                    collector[subject][ThrottledEmailBackend.THROTTLE_OVERALL] = this_overall
                    n_no += this_no
                    n_subject += this_subject
                    n_overall += this_overall

        return n_no, n_subject, n_overall, collector



django-email-throttler/management/commands/threshold_reporting.py
from optparse import make_option
import datetime
import dateutil.parser
import os
import glob

from django.core.mail import mail_admins
from django.core.management.base import BaseCommand
from django.conf import settings

from emailthrottler.throttler import ThrottledEmailBackend


class Command(BaseCommand):
    """
        Make a report about throttled mails and/or clean up the database.
        This command should be run once per throttle interval, preferably
        not much after the interval is over, most likely from cron.
        For example, if your interval is the default 600 seconds (10 minutes),
        then run it from cron as "*/10 * * * *" or so
    """

    help = "Report on email throttling"

    option_list = BaseCommand.option_list + (
        make_option(
            '--no-clean-log',
            action='store_true',
            dest='no_clean_log',
            default=False,
            help="Keep email log"
        ),
        make_option(
            '--force-email',
            action='store_true',
            dest='force_email',
            default=False,
            help="Send summary email even if there was no throttling (but only if there were emails in general)"
        ),
        make_option(
            '--no-email',
            action='store_true',
            dest='no_email',
            default=False,
            help="Don't sent summary email even if there would be something to report"
        ),
        make_option(
            "--start-date",
            action="store",
            dest="start_date",
            default=False,
            help="Use this as the starting time of an interval-to-check. Format is 'YYYY-MM-DD HH:MM' (UTC)"
        ),
    )

    # command handler
    def handle(self, *args, **options):

        # define our time interval
        if options['start_date']:
            bb = dateutil.parser.parse(options['start_date'])
        else:
            bb = self._bucket_begin()
        be = self._bucket_end(bb)

        # report if there are mails and either there's something to report or report is forced
        if not options['no_email']:
            self._send_report(bb, be, options['force_email'])

        if not options['no_clean_log']:
            self._clean_log(bb)

    # assemble and send email report
    @classmethod
    def _send_report(cls, bb, be, force_email):

        n_no, n_subject, n_overall, details = cls._collect_data(bb)

        subject = "Mail throttling report for {bb} -- T:{t} N:{n} S:{s} O:{o}".format(
            bb=bb.strftime('%Y-%m-%d %H:%M'),
            t=n_no+n_subject+n_overall,
            n=n_no,
            s=n_subject,
            o=n_overall,
        )

        out = ""
        out += "Report date: {bb} - {be}\n".format(bb=bb, be=be)
        out += "Total mails: {}\n".format(n_no+n_subject+n_overall)
        out += "Not throttled: {}\n".format(n_no)
        out += "Throttled on subject: {}\n".format(n_subject)
        out += "Throttled on overall amount: {}\n".format(n_overall)
        out += "\n"
        out += "T/N/S/O Subject\n"

        # dump the collection in a nice format
        for key, value in details.items():
            out += "{t}/{n}/{s}/{o}\t'{subject}'\n".format(
                t=value[ThrottledEmailBackend.THROTTLE_NO] +
                  value[ThrottledEmailBackend.THROTTLE_SUBJECT] +
                  value[ThrottledEmailBackend.THROTTLE_OVERALL],
                n=value[ThrottledEmailBackend.THROTTLE_NO],
                s=value[ThrottledEmailBackend.THROTTLE_SUBJECT],
                o=value[ThrottledEmailBackend.THROTTLE_OVERALL],
                subject=key)

        # send this report if there's something to say
        if (n_no+n_subject+n_overall > 0) and (force_email or n_subject > 0 or n_overall > 0):
            mail_admins(subject, out)

    # clean up the entries for the current interval from the db
    @classmethod
    def _clean_log(cls, bb):
        pattern = os.path.join(ThrottledEmailBackend.tmpdir_name,
                               ThrottledEmailBackend.file_prefix+str(bb)+"*")
        for f in glob.glob(pattern): 
            if os.path.isfile(f):
                os.remove(f)

    # assuming that the script is run not much after the current bucket started
    # calculate start time of the *previous* bucket
    @classmethod
    def _bucket_begin(cls):
        epoch_now = int(datetime.datetime.utcnow().strftime('%s'))
        return datetime.datetime.fromtimestamp(
            epoch_now - (epoch_now % ThrottledEmailBackend.interval_size) - ThrottledEmailBackend.interval_size
        )

    # calculate end time of a bucket based on the start time
    @classmethod
    def _bucket_end(cls, bb):
        return datetime.datetime.fromtimestamp(
            int(bb.strftime('%s')) + ThrottledEmailBackend.interval_size
        )

    # get number of total, subject/overall filtered messages for a bucket
    # and all details too
    @classmethod
    def _collect_data(cls, bb):
        n_no = 0
        n_subject = 0
        n_overall = 0
        collector = {}

        pattern = os.path.join(ThrottledEmailBackend.tmpdir_name,
                               ThrottledEmailBackend.file_prefix + str(bb) + "*")
        for f in glob.glob(pattern):
            if os.path.isfile(f):
                subject = f[f.index('^')+1:]

                collector[subject] = {}
                with open(os.path.join(ThrottledEmailBackend.tmpdir_name, f),
                          "r") as subject_file:
                    content = subject_file.read()

                    # count
                    this_no = content.count(str(ThrottledEmailBackend.THROTTLE_NO))
                    this_subject = content.count(str(ThrottledEmailBackend.THROTTLE_SUBJECT))
                    this_overall = content.count(str(ThrottledEmailBackend.THROTTLE_OVERALL))

                    # store per subject and update totals
                    collector[subject][ThrottledEmailBackend.THROTTLE_NO] = this_no
                    collector[subject][ThrottledEmailBackend.THROTTLE_SUBJECT] = this_subject
                    collector[subject][ThrottledEmailBackend.THROTTLE_OVERALL] = this_overall
                    n_no += this_no
                    n_subject += this_subject
                    n_overall += this_overall

        return n_no, n_subject, n_overall, collector



django_email_throttler/throttler.py
import datetime
import os
import glob
from base64 import urlsafe_b64encode

from django.core.mail.backends.base import BaseEmailBackend
from django.core import mail
from django.conf import settings


class ThrottledEmailBackend(BaseEmailBackend):

    THROTTLE_NO = 0
    THROTTLE_SUBJECT = 1
    THROTTLE_OVERALL = 2

    tmpdir_name = getattr(settings, 'EMAILTHROTTLER_TMPDIR', '/tmp')
    file_prefix = getattr(settings, 'EMAILTHROTTLER_PREFIX', 'emailthrottler-')
    interval_size = getattr(settings, 'EMAILTHROTTLER_INTERVAL', 600)
    subject_threshold = getattr(settings, 'EMAILTHROTTLER_SUBJECT_THRESHOLD', 3)
    overall_threshold = getattr(settings, 'EMAILTHROTTLER_OVERALL_THRESHOLD', 15)
    truncate_subject = getattr(settings, 'EMAILTHROTTLER_TRUNCATE_SUBJECT', None)

    # note: there's no open() an close()

    def send_messages(self, email_messages):

        bb = self._bucket_begin()
        in_bucket_total = self._get_in_bucket_sofar(bb)

        to_send = []
        for message in email_messages:

            # it'd be possible to cache this, but it's probably not worth it
            same_subj = self._get_same_subject_sofar(bb, message.subject)

            # check for overall throttling
            if self.overall_threshold>0 and \
                            in_bucket_total >= self.overall_threshold:
                throttled = self.THROTTLE_OVERALL
                #print("THROTTLE_OVERALL")

            # check for subject throttling
            elif self.subject_threshold>0 and \
                            same_subj >= self.subject_threshold:
                throttled = self.THROTTLE_SUBJECT
                #print("THROTTLE_SUBJECT")

            # seems to be below all thresholds, send it
            else:
                message.extra_headers['X-Mail-Throttler'] = (
                    'overall={overall}, nsubj={nsubj}'.format(
                        overall=in_bucket_total+1,
                        nsubj=same_subj+1
                    )
                )
                to_send.append(message)
                throttled = self.THROTTLE_NO
                #print("THROTTLE_NO")

            self._save_email_info(bb, message, throttled)
            in_bucket_total += 1

        connection = mail.get_connection()
        connection.open()
        connection.send_messages(to_send)
        connection.close()

    # calculate the beginning of the current bucket
    @classmethod
    def _bucket_begin(cls):
        epoch_now = int(datetime.datetime.utcnow().strftime('%s'))
        return datetime.datetime.fromtimestamp(
            epoch_now - (epoch_now % cls.interval_size)
        )

    # get total number of messages in the current bucket
    @classmethod
    def _get_in_bucket_sofar(cls, bb):
        total = 0
        pattern = os.path.join(cls.tmpdir_name,
                               cls.file_prefix+str(bb)+"*")
        for f in glob.glob(pattern): 
            if os.path.isfile(f):
                total += os.stat(f).st_size
        return total

    # get number messages in the current bucket with a given subject
    @classmethod
    def _get_same_subject_sofar(cls, bb, subject):
        try:
            filename = cls._get_file_name(bb, subject)
            return os.stat(filename).st_size
        except OSError:
            return 0

    # record meta info about this mail
    @classmethod
    def _save_email_info(cls, bb, message, throttled):
        filename = cls._get_file_name(bb, message.subject)
        with open(filename, "a") as out:
            out.write(str(throttled))

    # assemble the storage file name based on bucket begin + email subject
    @classmethod
    def _get_file_name(cls, bb, subject):
        if isinstance(subject, bytes):
            # Round trip to enforce ASCII
            subject_part = subject.decode("utf-8").encode("ascii")
        else:
            # Encode str as ASCII bytes
            subject_part = subject.encode("ascii")
        if cls.truncate_subject is not None:
            subject_part = subject_part[: cls.truncate_subject]
        return os.path.join(
            cls.tmpdir_name, cls.file_prefix + str(bb) + "^" +
            urlsafe_b64encode(subject_part).decode("ascii")
        )
